lib/spack/spack/solver/cache.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import hashlib
+import json
+import pathlib
+from contextlib import contextmanager
+from typing import List, Tuple, Union
+
+import llnl.util.tty as tty
+
+import spack
+import spack.config
+import spack.paths
+import spack.util.path
+from spack.solver.asp import Result
+
+
+class ConcretizationCache:
+    """Store for Spack concretization results and statistics
+
+    Serializes solver result objects and statistics to json and stores
+    at a given endpoint in a cache associated by the sha256 of the
+    asp problem and the involved control files.
+    """
+
+    def __init__(self, root: Union[str, None] = None):
+        if not root:
+            root = spack.config.get(
+                "config:concretization_cache", spack.paths.default_conc_cache_path
+            )
+        self.root = pathlib.Path(spack.util.path.canonicalize_path(root))
+        self.cleanup()
```

Comment:
I wouldn't run this in the constructor, but only on add.

Reply:
Agreed constructor is wrong place, just put it there because I couldn't decide where it should actually go. You think store is also right? I think both places are called to often...